### ISSUE TITLE
Potential loss of semantics in IO#to

### ISSUE THREAD

**Issue by djspiewak** [typelevel/cats-effect#993]
I'm not convinced this is sound:

```scala
            val poll2 = new (IO ~> IO) {
              def apply[B](ioa: IO[B]): IO[B] =
                F.to[IO](poll(ioa.to[F]))
            }
```

Specifically, the `F.to[IO]` bit. So the `poll` here is coming from `F.uncancelable`. Imagine, for example, if `F` is implementing `uncancelable` in the same fashion as `IO`, where there's a special constructor, `IO.Unmask`, which corresponds to it. This is not directly convertable, as we see a little later on in the function:

```scala
        case IO.Unmask(ioa, _) => ioa.to[F] // polling should be handled by F
```

When going from `IO` to `F`, we just ignore `Unmask`. So in other words, if you imagine a hypothetical `F` which has an internal implementation very much like `IO`'s, this function will not do the right thing! In fact, I'm relatively convinced that it doesn't do the right thing going from `IO` to `IO`, which ironically is law-tested, but the generators do not frequently create things which are sensitive to polling, and in any case our equality checkers cannot see it.

So tldr I think this is broken, and it *may* illustrate a broader brokenness with the concept of `Effect#to`. We almost certainly need to find a different way to implement this.

**Comment by djspiewak** [comment]
I'm pretty sure the solution here is to break `to` out into an inner function which carries along with it a `List[F ~> F]` which we treat as a stack: pushing/popping in `Uncanceled`, and peeking in `Unmask`. This isn't too hard to do, just requires a tiny bit of restructuring.

**Comment by vasilmkd** [comment]
@djspiewak I'm working on this, and here's what I've found so far.

First of all, your concern about `IO.Unmask` is valid, but I think there's another thing that has been overlooked. An `Unmask` constructor is only ever called in the `IO` run loop. Since `to[F]` doesn't actually run the `IO`, there are no composed `Unmask` `IO`s that need converting. You can verify this by replacing the `case IO.Unmask` with the following:

```scala
case IO.Unmask(_, _) => // ioa.to[F] // polling should be handled by F
          sys.error("Blah")
```

With that being said, I think I have an idea on how this can be solved. I'm exploring it.

**Comment by vasilmkd** [comment]
Addressed in #1036.
